ssl/ech.c

```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+
+#include <openssl/ssl.h>
+#include <openssl/ech.h>
+#include "ssl_local.h"
+#include "ech_local.h"
+#include "statem/statem_local.h"
+#include <openssl/rand.h>
+#include <openssl/trace.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+
+#ifndef OPENSSL_NO_ECH
+
+/* a size for some crypto vars */
+# define OSSL_ECH_CRYPTO_VAR_SIZE 2048
+
+/*
+ * @brief encode binary buffer as ascii hex
+ * @param out is an allocated buffer for the ascii hex string
+ * @param outsize is the size of the buffer
+ * @param in is the input binary buffer
+ * @param inlen is the size of the binary buffer
+ * @return 1 for good otherwise bad
+ */
+static int ah_encode(char *out, size_t outsize,
+                     const unsigned char *in, size_t inlen)
+{
+    size_t i;
+
+    if (outsize < 2 * inlen + 1)
+        return 0;
+    for (i = 0; i != inlen; i++) {
+        uint8_t tn = (in[i] >> 4) & 0x0f;
+        uint8_t bn = (in[i] & 0x0f);
+
+        out[2 * i] = (tn < 10 ? tn + '0' : (tn - 10 + 'A'));
+        out[2 * i + 1] = (bn < 10 ? bn + '0' : (bn - 10 + 'A'));
+    }
+    out[2 * i] = '\0';
+    return 1;
+}
+
+/*
+ * @brief has a buffer as a pretend file name being ascii-hex of hashed buffer
+ * @param buf is the input buffer
+ * @param blen is the length of buf
+ * @param ah_hash is a pointer to where to the result 
+ * @param ah_len is the length of ah_hash
+ */
+static int ech_hash_pub_as_fname(const unsigned char *buf, size_t blen,
+                                 char *ah_hash, size_t ah_len)
+{
+    EVP_MD *md = NULL;
+    EVP_MD_CTX *mdctx = NULL;
+    unsigned char hashval[EVP_MAX_MD_SIZE];
+    unsigned int hashlen;
+
+    if (((md = EVP_MD_fetch(NULL, "SHA2-256", NULL)) == NULL)
```

Comment:
Good point about libctx/propq - added that.

WRT fetching every time - I think that's ok as this will only be called when generating a new PEM file (via command line), or (in code for a later PR) when periodically (e.g. hourly) loading a key pair into haproxy via it's "stats" interface. This won't be needed when periodically loading keys into e.g. nginx from the filesystem. OTOH, we don't care at all what hash function is used for this, so if there were a good way to just get any old hash that'd be fine.